In [1]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report



In [2]:
data = pd.read_csv('Dataset .csv')

## Data Preprocessing

In [3]:
data.head()

Restaurant ID         Restaurant Name  Country Code              City  \
0        6317637        Le Petit Souffle           162       Makati City   
1        6304287        Izakaya Kikufuji           162       Makati City   
2        6300002  Heat - Edsa Shangri-La           162  Mandaluyong City   
3        6318506                    Ooma           162  Mandaluyong City   
4        6314302             Sambo Kojin           162  Mandaluyong City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2  Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3  Third Floor, Mega Fashion Hall, SM Megamall, O...   
4  Third Floor, Mega Atrium, SM Megamall, Ortigas...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   
2  Edsa Shangri-La, Ortigas, Mandaluyong City   
3      SM Megamall, Ortigas, Mandaluyong City   
4      SM Megamall, Ortigas, Mandaluyong City   

                                    Locality Verbose   Longitude   Latitude  \
0  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443   
1  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708   
2  Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831  14.581404   
3  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475  14.585318   
4  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508  14.584450   

                           Cuisines  ...          Currency Has Table booking  \
0        French, Japanese, Desserts  ...  Botswana Pula(P)               Yes   
1                          Japanese  ...  Botswana Pula(P)               Yes   
2  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)               Yes   
3                   Japanese, Sushi  ...  Botswana Pula(P)                No   
4                  Japanese, Korean  ...  Botswana Pula(P)               Yes   

  Has Online delivery Is delivering now Switch to order menu Price range  \
0                  No                No                   No           3   
1                  No                No                   No           3   
2                  No                No                   No           4   
3                  No                No                   No           4   
4                  No                No                   No           4   

   Aggregate rating  Rating color Rating text Votes  
0               4.8    Dark Green   Excellent   314  
1               4.5    Dark Green   Excellent   591  
2               4.4         Green   Very Good   270  
3               4.9    Dark Green   Excellent   365  
4               4.8    Dark Green   Excellent   229  

[5 rows x 21 columns]

In [4]:
data.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

In [5]:
np.array(data['Cuisines'])
data.iloc[:,9:10]

Cuisines
0           French, Japanese, Desserts
1                             Japanese
2     Seafood, Asian, Filipino, Indian
3                      Japanese, Sushi
4                     Japanese, Korean
...                                ...
9546                           Turkish
9547   World Cuisine, Patisserie, Cafe
9548            Italian, World Cuisine
9549                   Restaurant Cafe
9550                              Cafe

[9551 rows x 1 columns]

In [6]:
imputer = SimpleImputer(missing_values = np.nan, strategy='most_frequent')

imputer.fit(data.iloc[:,9:10].values)
data.iloc[:,9:10] = imputer.transform(data.iloc[:,9:10].values)
data.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                0
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

In [7]:
#dropping unnecessary columns
columns_to_drop = [
    'Restaurant ID', 'Country Code', 'City', 'Address', 'Locality', 
    'Locality Verbose', 'Longitude', 'Latitude', 'Currency', 
    'Has Table booking', 'Has Online delivery', 'Is delivering now', 
    'Switch to order menu', 'Price range', 'Aggregate rating', 
    'Rating color', 'Rating text', 'Votes'
]

data.drop(columns=columns_to_drop, axis=1, inplace=True)

In [8]:
data

Restaurant Name                          Cuisines  \
0             Le Petit Souffle        French, Japanese, Desserts   
1             Izakaya Kikufuji                          Japanese   
2       Heat - Edsa Shangri-La  Seafood, Asian, Filipino, Indian   
3                         Ooma                   Japanese, Sushi   
4                  Sambo Kojin                  Japanese, Korean   
...                        ...                               ...   
9546               Naml۱ Gurme                           Turkish   
9547              Ceviz A��ac۱   World Cuisine, Patisserie, Cafe   
9548                     Huqqa            Italian, World Cuisine   
9549               A���k Kahve                   Restaurant Cafe   
9550  Walter's Coffee Roastery                              Cafe   

      Average Cost for two  
0                     1100  
1                     1200  
2                     4000  
3                     1500  
4                     1500  
...                    ...  
9546                    80  
9547                   105  
9548                   170  
9549                   120  
9550                    55  

[9551 rows x 3 columns]

In [9]:
#Converting Categorical Data to Numerical
label_encoder = LabelEncoder() 
data['Restaurant Name'] = label_encoder.fit_transform(data['Restaurant Name']) 
data['Cuisines'] = label_encoder.fit_transform(data['Cuisines'])

In [10]:
data

Restaurant Name  Cuisines  Average Cost for two
0                3748       920                  1100
1                3172      1111                  1200
2                2896      1671                  4000
3                4707      1126                  1500
4                5523      1122                  1500
...               ...       ...                   ...
9546             4443      1813                    80
9547             1310      1824                   105
9548             3068      1110                   170
9549              512      1657                   120
9550             7240       331                    55

[9551 rows x 3 columns]

## Building Random Forest Model


In [11]:
label_encoder_restaurant = LabelEncoder()
label_encoder_cuisines = LabelEncoder()
data['Restaurant Name'] = label_encoder_restaurant.fit_transform(data['Restaurant Name'])
data['Cuisines'] = label_encoder_cuisines.fit_transform(data['Cuisines'])


In [12]:
X = data[['Restaurant Name', 'Average Cost for two']]
Y = data['Cuisines']

In [13]:
scaler = StandardScaler()
X = scaler.fit_transform(X)
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=10)


In [14]:
model_DT = DecisionTreeClassifier(random_state=10, criterion="gini")
model_DT.fit(X_train, Y_train)
Y_pred = model_DT.predict(X_test)

In [15]:
print("Actual vs Predicted:")
print(list(zip(Y_test, Y_pred)))



Actual vs Predicted:
[(1287, 1514), (517, 1782), (1306, 1306), (497, 1306), (497, 582), (1329, 1182), (1699, 1329), (1306, 1329), (186, 1286), (849, 201), (1306, 518), (1306, 1329), (1306, 1306), (1329, 1348), (982, 774), (913, 913), (982, 1306), (1406, 560), (1789, 1789), (1514, 1154), (1373, 831), (58, 58), (1514, 1329), (1546, 1329), (828, 828), (1406, 1514), (1329, 1329), (58, 58), (732, 637), (497, 1329), (895, 790), (1448, 556), (1514, 1288), (1111, 1293), (1111, 625), (186, 201), (910, 697), (533, 588), (1048, 797), (417, 252), (1749, 758), (518, 518), (1554, 310), (1329, 1699), (1159, 251), (1618, 1618), (933, 154), (828, 868), (895, 1686), (828, 549), (22, 270), (1306, 765), (1435, 628), (1231, 168), (1626, 1626), (1589, 1111), (177, 1394), (1749, 1750), (1306, 1306), (1306, 1590), (1520, 1514), (828, 828), (1288, 177), (1769, 1306), (1124, 1126), (778, 1667), (412, 1401), (583, 1235), (1618, 1618), (1329, 1122), (549, 595), (1329, 1514), (856, 251), (1288, 1288), (1655, 1655)

In [16]:
# Evaluating the model
print("\nClassification Report:")
print(classification_report(Y_test, Y_pred))



Classification Report:
              precision    recall  f1-score   support

           6       0.40      0.40      0.40         5
           7       0.00      0.00      0.00         1
           9       0.00      0.00      0.00         1
          11       0.00      0.00      0.00         1
          12       0.00      0.00      0.00         0
          15       0.00      0.00      0.00         1
          16       0.00      0.00      0.00         0
          18       0.00      0.00      0.00         2
          21       0.00      0.00      0.00         0
          22       0.00      0.00      0.00         1
          26       0.00      0.00      0.00         0
          29       0.00      0.00      0.00         2
          34       0.00      0.00      0.00         1
          35       0.00      0.00      0.00         0
          36       0.00      0.00      0.00         1
          42       0.00      0.00      0.00         1
          45       0.00      0.00      0.00         0
   

C:\Users\Sakshi\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Sakshi\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Sakshi\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Sakshi\anaconda3\lib\site-packages\sklearn\metrics\

In [17]:
print("Accuracy Score:", accuracy_score(Y_test, Y_pred))

Accuracy Score: 0.24071166928309787


## Building Logistic Regression

In [19]:
classifier = LogisticRegression(multi_class="multinomial")
#train the model object
classifier.fit(X_train,Y_train)      # fit is the function that is used for training the data

C:\Users\Sakshi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(multi_class='multinomial')

In [20]:
Y_pred = classifier.predict(X_test)
print(Y_pred)

[1306 1306 1306 ... 1306 1306 1306]


In [26]:
cf_m = confusion_matrix(Y_test, Y_pred)
accuracy = accuracy_score(Y_test, Y_pred)
class_report = classification_report(Y_test, Y_pred)

C:\Users\Sakshi\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Sakshi\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Sakshi\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [23]:
print("\nConfusion Matrix:")
print(cf_m)


Confusion Matrix:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [24]:
print("\nClassification Report:")
print(class_report)


Classification Report:
              precision    recall  f1-score   support

           6       0.00      0.00      0.00         5
           7       0.00      0.00      0.00         1
           9       0.00      0.00      0.00         1
          11       0.00      0.00      0.00         1
          15       0.00      0.00      0.00         1
          18       0.00      0.00      0.00         2
          22       0.00      0.00      0.00         1
          29       0.00      0.00      0.00         2
          34       0.00      0.00      0.00         1
          36       0.00      0.00      0.00         1
          42       0.00      0.00      0.00         1
          46       0.00      0.00      0.00         1
          54       0.00      0.00      0.00         5
          58       0.00      0.00      0.00         7
          61       0.00      0.00      0.00         1
          69       0.00      0.00      0.00         2
          74       0.00      0.00      0.00         1
   

In [28]:
print("\nAccuracy of the model:")
print(accuracy)


Accuracy of the model:
0.09419152276295134


## Conclusion

When comparing the two models, Random Forest demonstrates superior performance with an accuracy of 0.24071166928309787, compared to Logistic Regression's accuracy of 0.09419152276295134. This indicates that Random Forest is more effective at predicting the cuisines in this dataset, making it the better choice for this classification task.